### PR TITLE
fixes #9409 feat(nimbus): Vendor FML files for all FML-based applications

### LINKED ISSUE #9409: Vendor FML files and use them to generate experimenter yaml files
We can use nimbus-cli to fetch the single-file FML files based on the apps listed in apps.yaml, and then generate experimenter.yaml files. This will also help with the FML-in-experimenter work.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3836)

### PR DESCRIPTION
**Pull request by brennie**
Because
- we want to use FML in experimenter

this commit
- adds nimbus-cli to the experimenter:dev docker image;
- adds a wrapper around nimbus-cli (manifest-tool) for fetching all the
  FML files and generating experimenter.yaml files;
- updates apps.yaml to include monitor; and
- does an initial import of all the FML files.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -89,11 +89,7 @@ jetstream_config:
 feature_manifests:
 	mkdir -p $(MANIFESTS_DIR)
 
-	$(NIMBUS_CLI) fml -- generate-experimenter --channel release $(FEATURE_MANIFEST_FENIX) "$(MANIFESTS_DIR)/fenix/experimenter.yaml"
-	$(NIMBUS_CLI) fml -- generate-experimenter --channel release $(FEATURE_MANIFEST_FXIOS) "$(MANIFESTS_DIR)/ios/experimenter.yaml"
-	$(NIMBUS_CLI) fml -- generate-experimenter --channel release $(FEATURE_MANIFEST_FOCUS_ANDROID) "$(MANIFESTS_DIR)/focus-android/experimenter.yaml"
-	$(NIMBUS_CLI) fml -- generate-experimenter --channel release $(FEATURE_MANIFEST_FOCUS_IOS) "$(MANIFESTS_DIR)/focus-ios/experimenter.yaml"
-	$(NIMBUS_CLI) fml -- generate-experimenter --channel production $(FEATURE_MANIFEST_MONITOR) "$(MANIFESTS_DIR)/monitor-web/experimenter.yaml"
+	$(COMPOSE) run experimenter /experimenter/bin/manifest-tool.py fetch-latest
```
Comment (jaredlockhart) [inline Makefile:92]: 😲

File: experimenter/bin/manifest-tool.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env python
+
+import sys
+from pathlib import Path
+
+
+if __name__ == "__main__":
+    sys.path.append(str(Path(__file__).parent.parent / "experimenter" / "features"))
+
+    from manifesttool import main
```
Comment (jaredlockhart) [inline experimenter/bin/manifest-tool.py:10]: Why is this necessary?

File: experimenter/experimenter/features/manifests/apps.yaml [hunk 3]
```diff
@@ -3,22 +3,35 @@
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 ---
 fenix:
-  repo: \@mozilla-mobile/firefox-android
-  fml_path: fenix/app/nimbus.fml.yaml
+  slug: fenix
+  repo: "mozilla-mobile/firefox-android"
+  fml_path: "fenix/app/nimbus.fml.yaml"
+  channels: ["release", "beta", "nightly", "developer"]
```
Comment (jaredlockhart) [inline experimenter/experimenter/features/manifests/apps.yaml:9]: We can't discover these from the FMLs?

File: experimenter/experimenter/features/manifesttool/__init__.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        for channel in app_config.channels:
+            nimbus_cli(
+                "fml",
+                "--",
+                "single-file",
+                "--channel",
+                channel,
+                f"@{app_config.repo}/{app_config.fml_path}",
+                str(MANIFESTS_DIR / app_config.slug / f"{channel}.fml.yaml"),
+            )
```
Comment (jaredlockhart) [inline experimenter/experimenter/features/manifesttool/__init__.py:50]: Huh, never seen `click` before.  This is fancy.

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
This is so super cool 🎉 🎉 🎉 

But I'm getting a build error locally

```
 => ERROR [dev 43/43] RUN mkdir -p /application-services/bin/ &&     curl --proto '=https' --tlsv1.2 -sSf -o "/application-services/install-nimbus-cli.sh" "https://raw  0.3s
------
 > [dev 43/43] RUN mkdir -p /application-services/bin/ &&     curl --proto '=https' --tlsv1.2 -sSf -o "/application-services/install-nimbus-cli.sh" "https://raw.githubusercontent.com/mozilla/application-services/main/install-nimbus-cli.sh" &&     bash /application-services/install-nimbus-cli.sh --directory /application-services/bin:
0.104 ✅ Downloading https://firefox-ci-tc.services.mozilla.com/api/index/v1/task/project.application-services.v2.nimbus-cli.latest/artifacts/public/build/nimbus-cli-aarch64-unknown-linux-gnu.zip
0.283 ✅ Installing into /application-services/bin
0.286 [/tmp/nimbus-cli-installer-ZcKZXH/nimbus-cli-aarch64-unknown-linux-gnu.zip]
0.286   End-of-central-directory signature not found.  Either this file is not
0.286   a zipfile, or it constitutes one disk of a multi-part archive.  In the
0.286   latter case the central directory and zipfile comment will be found on
0.286   the last disk(s) of this archive.
0.286 unzip:  cannot find zipfile directory in one of /tmp/nimbus-cli-installer-ZcKZXH/nimbus-cli-aarch64-unknown-linux-gnu.zip or
0.286         /tmp/nimbus-cli-installer-ZcKZXH/nimbus-cli-aarch64-unknown-linux-gnu.zip.zip, and cannot find /tmp/nimbus-cli-installer-ZcKZXH/nimbus-cli-aarch64-unknown-linux-gnu.zip.ZIP, period.
0.286 ✅ Cleaning up installation directory
0.287 ❎ Failed to install nimbus-cli with the arguments provided: --directory /application-services/bin, at line number 1
0.287   For support, go to https://experimenter.info/nimbus-cli/install
------
ERROR: failed to solve: process "/bin/sh -c mkdir -p /application-services/bin/ &&     curl --proto '=https' --tlsv1.2 -sSf -o \"/application-services/install-nimbus-cli.sh\" \"https://raw.githubusercontent.com/mozilla/application-services/main/install-nimbus-cli.sh\" &&     bash /application-services/install-nimbus-cli.sh --directory /application-services/bin" did not complete successfully: exit code: 9

View build details: docker-desktop://dashboard/build/cloud-mozilla-default/cloud-mozilla-default0/occ2ahnjpdwmm7ta00h0dkzn3
make: *** [build_dev] Error 1
```


Throwing a `cat` into the install script tells me

```
root@3c1541903bed:/experimenter# bash /application-services/install-nimbus-cli.sh --directory /application-services/bin
✅ Downloading https://firefox-ci-tc.services.mozilla.com/api/index/v1/task/project.application-services.v2.nimbus-cli.latest/artifacts/public/build/nimbus-cli-aarch64-unknown-linux-gnu.zip
✅ Installing into /application-services/bin
{
  "code": "ResourceNotFound",
  "message": "Artifact not found\n\n---\n\n* method:     getLatestArtifact\n* errorCode:  ResourceNotFound\n* statusCode: 404\n* time:       2023-09-21T18:51:21.410Z",
  "requestInfo": {
    "method": "getLatestArtifact",
    "params": {
      "0": "public/build/nimbus-cli-aarch64-unknown-linux-gnu.zip",
      "taskId": "X--t4Dd9Q1OYf5_I8dqsBg",
      "name": "public/build/nimbus-cli-aarch64-unknown-linux-gnu.zip"
    },
    "payload": {},
    "time": "2023-09-21T18:51:21.410Z"
  }
}[/tmp/nimbus-cli-installer-imR3QX/nimbus-cli-aarch64-unknown-linux-gnu.zip]
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of /tmp/nimbus-cli-installer-imR3QX/nimbus-cli-aarch64-unknown-linux-gnu.zip or
        /tmp/nimbus-cli-installer-imR3QX/nimbus-cli-aarch64-unknown-linux-gnu.zip.zip, and cannot find /tmp/nimbus-cli-installer-imR3QX/nimbus-cli-aarch64-unknown-linux-gnu.zip.ZIP, period.
✅ Cleaning up installation directory
❎ Failed to install nimbus-cli with the arguments provided: --directory /application-services/bin, at line number 1
        For support, go to https://experimenter.info/nimbus-cli/install
```

File: experimenter/bin/manifest-tool.py (hunk 2 above)
Comment (brennie) [reply experimenter/bin/manifest-tool.py:10]: It lets you import manifesttool directory instead of importing `experimeter.features.manifesttool`. Because the experimenter package is a Django app, there are side-effects from importing `experimenter` directly. Basically lets us have a CLI app without importing initializing all the django gunk

File: experimenter/experimenter/features/manifests/apps.yaml (hunk 3 above)
Comment (brennie) [reply experimenter/experimenter/features/manifests/apps.yaml:9]: This is temporary to get things working. We either have to add code here to read and parse the remote FML file to discover channels *or* add support to nimbus-cli to discover channels or add an --all-channels flag to single-file command.

Right now we need to specify a channel for single-file and the resulting single file yaml only has one channel listed.

**Comment by brennie** [comment]
> This is so super cool 🎉 🎉 🎉
> 
> But I'm getting a build error locally
> 

Apparently the artifacts are not rebuilt after each PR, but nightly. We need to wait until tomorrow for it to be available, so I'm marking this PR blocked until then.

**Comment by brennie** [comment]
This is no longer blocked: the [index](https://firefox-ci-tc.services.mozilla.com/tasks/index/project.application-services.v2.nimbus-cli/latest) has updated with the new binaries

File: experimenter/.coveragerc [hunk 5]
```diff
@@ -8,6 +8,7 @@ omit =
 [report]
 fail_under = 100
 omit =
+    bin/manifest-tool.py
```
Comment (jaredlockhart) [inline experimenter/.coveragerc:11]: Are we still gonna skip coverage checking this whole file?

File: experimenter/bin/manifest-tool.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env python
+
+import sys
+from pathlib import Path
+
+
+if __name__ == "__main__":
+    # sys.path.append(str(Path(__file__).parent.parent / "experimenter" / "features"))
```
Comment (jaredlockhart) [inline experimenter/bin/manifest-tool.py:8]: 👋

File: experimenter/experimenter/features/manifests/apps.yaml (hunk 3 above)
Comment (jaredlockhart) [reply experimenter/experimenter/features/manifests/apps.yaml:9]: Perfect 👍

File: experimenter/manifesttool/tests/test_cli.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+class FetchLatestTests(TestCase):
+    """Tests for the fetch-latest subcommand."""
+
+    @patch.object(cli, "download_single_file", mock_download_single_file)
+    def test_valid_manifest(self):
+        """Test with a valid apps.yaml"""
+
+        runner = CliRunner()
+        with runner.isolated_filesystem():
```
Comment (jaredlockhart) [inline experimenter/manifesttool/tests/test_cli.py:72]: Wow neat

File: experimenter/.coveragerc (hunk 5 above)
Comment (jaredlockhart) [reply experimenter/.coveragerc:11]: Oh I see, are we invoking the test below outside of the full Experimenter suite?

File: experimenter/manifesttool/tests/test_cli.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+from click.testing import CliRunner
+import yaml
+
+from manifesttool import cli
+from manifesttool.appconfig import AppConfig, AppConfigs
+
+
+# An AppConfig with a blank repo and fml_path so that the CLI cannot fetch
+# anything.
```
Comment (jaredlockhart) [inline experimenter/manifesttool/tests/test_cli.py:13]: Ah okay so this tests that we can parse the app config file, trigger the cli in a networkless no op mode, and then write out and exit successfully.  Cool.

Comment (jaredlockhart) [reply experimenter/manifesttool/tests/test_cli.py:13]: Where does this test get invoked?

**Review by jaredlockhart — COMMENTED** [review]
Cool yeah this is looking great 🎉

File: experimenter/.coveragerc (hunk 5 above)
Comment (brennie) [reply experimenter/.coveragerc:11]: All bin/manifest-tool.py does is call main(). We test the cli in test_cli.py

File: experimenter/manifesttool/tests/test_cli.py (hunk 8 above)
Comment (brennie) [reply experimenter/manifesttool/tests/test_cli.py:13]: It gets invoked by pytest

Comment (jaredlockhart) [reply experimenter/manifesttool/tests/test_cli.py:13]: Oh pytest just discovers and runs it.  Well so much for my plan to deprecate it.  Cool 👍

Comment (brennie) [reply experimenter/manifesttool/tests/test_cli.py:13]: You can see it in the test run [here](https://app.circleci.com/pipelines/github/mozilla/experimenter/29340/workflows/d678a55e-11e2-4e94-8d7a-31f2263b9679/jobs/112000/parallel-runs/0/steps/0-104). My next push will include it in coverage, but my local test run shows 100% coverage:
```
manifesttool/tests/test_cli.py::FetchLatestTests::test_valid_manifest PASSED                                                                                                                              [100%]

---------- coverage: platform linux, python 3.11.1-final-0 -----------
Name                                                                       Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------------------------
manifesttool/__init__.py                                                       0      0   100%
manifesttool/appconfig.py                                                     18      0   100%
manifesttool/cli.py                                                           26      0   100%
manifesttool/nimbus_cli.py                                                    10      0   100%
```

**Review by jaredlockhart — APPROVED** [review]
Super cool, looks great, thanks for adding tests 🎉 🎉 🎉 🎉 🎉

**Comment by yashikakhurana** [comment]
Just wanted to make sure, that FML syncs with the repository of the implementing application, this still should work? but now if its both for staging and production, lets say monitor, it should update two files?
